codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrame.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+/**
+ * A frame whose meaning <em>may</em> apply to a particular stream, instead of the entire
+ * connection. It is still possibly for this frame type to apply to the entire connection. In such
+ * cases, the {@code stream} reference should be {@code null} or an object referring to the
+ * connection.
+ *
+ * <p>The meaning of {@code stream} is context-dependent and may change as a frame is processed in
+ * the pipeline.
+ */
+public interface Http2StreamFrame extends Http2Frame {
+    /**
+     * Set the stream identifier for this message.
+     *
+     * @return {@code this}
+     */
+    Http2StreamFrame setStream(Object stream);
```

Comment:
> Right now my plan is to change the value over time for a particular message
> This is part of a larger design... Once I do the split, I think we'd be in a much better position to discuss alternative designs for storing this value.

OK. My suggestion was based upon the current code which appears to only use this before writing a frame.

> In the case of having a factory for HttpFrame objects, that seems to completely prevent any extending/custom implementations....Making the channel final would be counter-productive for my objectives. I'm trying to provide a mechanism that would allow unknown HTTP/2 frame type pass-through and prevent allocations/copies.

Not necessarily. The factory could be pluggable / extensible and conditional based upon some sort of frame type.  This is just a suggestion...more just a way to explore the potential benefits/pitfalls of having an immutable interface, and maybe understand a bit more as to how you plan to expose the Http2StreamChannel to the user. The biggest limitation of the "factory" would be potentially the "tight" coupling for object creation, increase dependency between classes, user would have to be aware of this class outside the pipeline.

> We already have a clear idea of the lifetime of the created stream. But even so, a factory wouldn't tell us anything about lifetime, because objects can be created long before they reach the head of the pipeline.

The "lifetime" reference is such that when the frame is created the channel would be known (or could be created).  There will need to be some sort of trigger point for this because if a user is going to initiate stream creation with a write operation, won't they will need the stream channel?